File: apps/android/app/src/main/kotlin/dev/hyphen/android/transport/Envelope.kt
package dev.hyphen.android.transport

/**
 * Protocol v0 envelope (HYP-M2-012, protocol doc §3) — model + strict
 * codec matching `protocol/schema/envelope.schema.json` exactly: unknown
 * fields rejected, patterns enforced, `sessionId` nullable (null only in
 * hello per session logic — structurally nullable here, like the schema).
 */
data class Envelope(
    val protocol: String = PROTOCOL_ID,
    val messageId: String,
    val sessionId: String?,
    val type: String,
    val capability: String? = null,
    val seq: Long,
    val ackOf: String? = null,
    val sentAtUnixMs: Long,
    val requiresAck: Boolean,
    val payload: Json.Obj = Json.Obj.EMPTY,
    val trace: Json.Obj? = null,
) {
    fun encode(): ByteArray {
        val entries = LinkedHashMap<String, Json>()
        entries["protocol"] = Json.Str(protocol)
        entries["messageId"] = Json.Str(messageId)
        entries["sessionId"] = sessionId?.let { Json.Str(it) } ?: Json.Null
        entries["type"] = Json.Str(type)
        capability?.let { entries["capability"] = Json.Str(it) }
        entries["seq"] = Json.Num(seq.toString())
        ackOf?.let { entries["ackOf"] = Json.Str(it) }
        entries["sentAtUnixMs"] = Json.Num(sentAtUnixMs.toString())
        entries["requiresAck"] = Json.Bool(requiresAck)
        entries["payload"] = payload
        trace?.let { entries["trace"] = it }
        return Json.Obj(entries).encode().toByteArray(Charsets.UTF_8)
    }

    companion object {
        /** Protocol identifier this implementation speaks. */
        const val PROTOCOL_ID = "hyphen/0.3"

        const val TYPE_HEARTBEAT = "heartbeat"
        const val TYPE_ACK = "ack"
        const val TYPE_HELLO = "hello"
        const val TYPE_ERROR = "error"

        private val KNOWN_FIELDS = setOf(
            "protocol", "messageId", "sessionId", "type", "capability",
            "seq", "ackOf", "sentAtUnixMs", "requiresAck", "payload", "trace",
        )
        private val PROTOCOL_SHAPE = Regex("^hyphen/0\\.[0-9]+$")
        private val TYPE_SHAPE = Regex("^[a-z][a-z0-9]*(\\.[a-z][a-z0-9]*)*$")
        private val SESSION_SHAPE = Regex("^s_[A-Za-z0-9_-]{2,64}$")
        private val CAPABILITY_SHAPE = Regex("^[a-z][a-z0-9]*\\.v[0-9]+$")

        fun decode(bytes: ByteArray): Envelope {
            val root = try {
                Json.parse(bytes.toString(Charsets.UTF_8))
            } catch (e: JsonParseException) {
                throw EnvelopeException("not JSON: ${e.message}")
            }
            val obj = root as? Json.Obj ?: throw EnvelopeException("envelope must be an object")

            val unknown = obj.entries.keys - KNOWN_FIELDS
            if (unknown.isNotEmpty()) throw EnvelopeException("unknown fields: $unknown")

            val protocol = string(obj, "protocol")
            if (!PROTOCOL_SHAPE.matches(protocol)) throw EnvelopeException("bad protocol '$protocol'")

            val messageId = string(obj, "messageId")
            if (!Ulid.isValid(messageId)) throw EnvelopeException("messageId is not a ULID")

            val sessionId = when (val raw = obj["sessionId"] ?: throw EnvelopeException("sessionId missing")) {
                is Json.Null -> null
                is Json.Str ->
                    raw.value.takeIf { SESSION_SHAPE.matches(it) }
                        ?: throw EnvelopeException("bad sessionId")
                else -> throw EnvelopeException("sessionId must be string or null")
            }

            val type = string(obj, "type")
            if (!TYPE_SHAPE.matches(type)) throw EnvelopeException("bad type '$type'")

            val capability = optionalString(obj, "capability")?.also {
                if (!CAPABILITY_SHAPE.matches(it)) throw EnvelopeException("bad capability '$it'")
            }

            val seq = long(obj, "seq")
            if (seq < 1) throw EnvelopeException("seq must be >= 1")

            val ackOf = when (val raw = obj["ackOf"]) {
                null, is Json.Null -> null
                is Json.Str ->
                    raw.value.takeIf(Ulid::isValid) ?: throw EnvelopeException("ackOf is not a ULID")
                else -> throw EnvelopeException("ackOf must be string or null")
            }

            val sentAtUnixMs = long(obj, "sentAtUnixMs")
            if (sentAtUnixMs < 0) throw EnvelopeException("sentAtUnixMs must be >= 0")

            val requiresAck = (obj["requiresAck"] as? Json.Bool
                ?: throw EnvelopeException("requiresAck missing or not boolean")).value

            val payload = obj["payload"] as? Json.Obj
                ?: throw EnvelopeException("payload missing or not an object")

            val trace = when (val raw = obj["trace"]) {
                null -> null
                is Json.Obj -> validateTrace(raw)
                else -> throw EnvelopeException("trace must be an object")
            }

            return Envelope(
                protocol, messageId, sessionId, type, capability,
                seq, ackOf, sentAtUnixMs, requiresAck, payload, trace,
            )
        }

        private fun validateTrace(trace: Json.Obj): Json.Obj {
            val unknown = trace.entries.keys - setOf("localOnly", "spanId")
            if (unknown.isNotEmpty()) throw EnvelopeException("unknown trace fields: $unknown")
            val localOnly = trace["localOnly"] as? Json.Bool
                ?: throw EnvelopeException("trace.localOnly missing or not boolean")
            if (!localOnly.value) throw EnvelopeException("trace.localOnly must be true")
            when (val spanId = trace["spanId"]) {
                null -> Unit
                is Json.Str -> {
                    if (!ProtocolTrace.isValidSpanId(spanId.value)) {
                        throw EnvelopeException("trace.spanId must be a ULID")
                    }
                }
                else -> throw EnvelopeException("trace.spanId must be a string")
            }
            return trace
        }

        private fun string(obj: Json.Obj, field: String): String =
            (obj[field] as? Json.Str ?: throw EnvelopeException("$field missing or not a string")).value

        private fun optionalString(obj: Json.Obj, field: String): String? =
            when (val raw = obj[field]) {
                null -> null
                is Json.Str -> raw.value
                else -> throw EnvelopeException("$field must be a string")
            }

        private fun long(obj: Json.Obj, field: String): Long =
            (obj[field] as? Json.Num ?: throw EnvelopeException("$field missing or not a number"))
                .asLong() ?: throw EnvelopeException("$field is not an integer")
    }
}

class ProtocolTrace private constructor(val spanId: String) {
    fun toJson(): Json.Obj =
        Json.obj(
            "localOnly" to Json.Bool(true),
            "spanId" to Json.Str(spanId),
        )

    companion object {
        fun local(spanId: String = Ulid.generate()): ProtocolTrace {
            require(isValidSpanId(spanId)) { "spanId must be a ULID" }
            return ProtocolTrace(spanId)
        }

        fun isValidSpanId(spanId: String): Boolean = Ulid.isValid(spanId)
    }
}

/** Maps to `protocol/invalid-envelope` in the error taxonomy. */
class EnvelopeException(message: String) : Exception(message)

File: apps/macos/Sources/HyphenTransport/Envelope.swift
import Foundation
import HyphenCore

/// Maps to `protocol/invalid-envelope` in the error taxonomy.
public struct EnvelopeError: Error, CustomStringConvertible {
    public let detail: String
    public var description: String { detail }
    init(_ detail: String) { self.detail = detail }
}

public struct ProtocolTrace: Equatable {
    public let spanId: String

    private init(spanId: String) {
        self.spanId = spanId
    }

    public static func local(spanId: String = Ulid.generate()) throws -> ProtocolTrace {
        guard isValidSpanId(spanId) else { throw EnvelopeError("trace.spanId must be a ULID") }
        return ProtocolTrace(spanId: spanId)
    }

    public static func isValidSpanId(_ spanId: String) -> Bool {
        Ulid.isValid(spanId)
    }

    public var wireObject: [String: Any] {
        [
            "localOnly": true,
            "spanId": spanId,
        ]
    }
}

/// Protocol v0 envelope (HYP-M2-012, protocol doc §3) — model + strict
/// codec matching `protocol/schema/envelope.schema.json` exactly: unknown
/// fields rejected, patterns enforced, `sessionId` nullable. JSON via
/// Foundation's JSONSerialization (payload stays an arbitrary object).
public struct Envelope {

    public static let typeHeartbeat = "heartbeat"
    public static let typeAck = "ack"
    public static let typeHello = "hello"
    public static let typeError = "error"

    public let protocolId: String
    public let messageId: String
    public let sessionId: String?
    public let type: String
    public let capability: String?
    public let seq: Int64
    public let ackOf: String?
    public let sentAtUnixMs: Int64
    public let requiresAck: Bool
    public let payload: [String: Any]
    public let trace: [String: Any]?

    public init(
        protocolId: String = HyphenCore.protocolVersion,
        messageId: String,
        sessionId: String?,
        type: String,
        capability: String? = nil,
        seq: Int64,
        ackOf: String? = nil,
        sentAtUnixMs: Int64,
        requiresAck: Bool,
        payload: [String: Any] = [:],
        trace: [String: Any]? = nil
    ) {
        self.protocolId = protocolId
        self.messageId = messageId
        self.sessionId = sessionId
        self.type = type
        self.capability = capability
        self.seq = seq
        self.ackOf = ackOf
        self.sentAtUnixMs = sentAtUnixMs
        self.requiresAck = requiresAck
        self.payload = payload
        self.trace = trace
    }

    public func encode() throws -> Data {
        var dict: [String: Any] = [
            "protocol": protocolId,
            "messageId": messageId,
            "sessionId": sessionId ?? NSNull(),
            "type": type,
            "seq": seq,
            "sentAtUnixMs": sentAtUnixMs,
            "requiresAck": requiresAck,
            "payload": payload,
        ]
        if let capability { dict["capability"] = capability }
        if let ackOf { dict["ackOf"] = ackOf }
        if let trace { dict["trace"] = trace }
        return try JSONSerialization.data(withJSONObject: dict, options: [.sortedKeys])
    }

    private static let knownFields: Set<String> = [
        "protocol", "messageId", "sessionId", "type", "capability",
        "seq", "ackOf", "sentAtUnixMs", "requiresAck", "payload", "trace",
    ]

    public static func decode(_ data: Data) throws -> Envelope {
        let root: Any
        do {
            root = try JSONSerialization.jsonObject(with: data, options: [])
        } catch {
            throw EnvelopeError("not JSON: \(error.localizedDescription)")
        }
        guard let obj = root as? [String: Any] else {
            throw EnvelopeError("envelope must be an object")
        }

        let unknown = Set(obj.keys).subtracting(knownFields)
        guard unknown.isEmpty else { throw EnvelopeError("unknown fields: \(unknown.sorted())") }

        let protocolId = try string(obj, "protocol", pattern: "^hyphen/0\\.[0-9]+$")
        let messageId = try string(obj, "messageId")
        guard Ulid.isValid(messageId) else { throw EnvelopeError("messageId is not a ULID") }

        guard let rawSession = obj["sessionId"] else { throw EnvelopeError("sessionId missing") }
        let sessionId: String?
        if rawSession is NSNull {
            sessionId = nil
        } else if let value = rawSession as? String,
                  value.range(of: "^s_[A-Za-z0-9_-]{2,64}$", options: .regularExpression) != nil {
            sessionId = value
        } else {
            throw EnvelopeError("bad sessionId")
        }

        let type = try string(obj, "type", pattern: "^[a-z][a-z0-9]*(\\.[a-z][a-z0-9]*)*$")

        var capability: String?
        if obj["capability"] != nil {
            capability = try string(obj, "capability", pattern: "^[a-z][a-z0-9]*\\.v[0-9]+$")
        }

        let seq = try integer(obj, "seq")
        guard seq >= 1 else { throw EnvelopeError("seq must be >= 1") }

        var ackOf: String?
        if let rawAckOf = obj["ackOf"], !(rawAckOf is NSNull) {
            guard let value = rawAckOf as? String, Ulid.isValid(value) else {
                throw EnvelopeError("ackOf is not a ULID")
            }
            ackOf = value
        }

        let sentAtUnixMs = try integer(obj, "sentAtUnixMs")
        guard sentAtUnixMs >= 0 else { throw EnvelopeError("sentAtUnixMs must be >= 0") }

        guard let rawAck = obj["requiresAck"], isBoolean(rawAck), let requiresAck = rawAck as? Bool else {
            throw EnvelopeError("requiresAck missing or not boolean")
        }

        guard let payload = obj["payload"] as? [String: Any] else {
            throw EnvelopeError("payload missing or not an object")
        }

        var trace: [String: Any]?
        if let rawTrace = obj["trace"] {
            guard let traceObj = rawTrace as? [String: Any] else {
                throw EnvelopeError("trace must be an object")
            }
            let extras = Set(traceObj.keys).subtracting(["localOnly", "spanId"])
            guard extras.isEmpty else { throw EnvelopeError("unknown trace fields: \(extras.sorted())") }
            guard let rawLocalOnly = traceObj["localOnly"], isBoolean(rawLocalOnly), let localOnly = rawLocalOnly as? Bool else {
                throw EnvelopeError("trace.localOnly missing or not boolean")
            }
            guard localOnly else { throw EnvelopeError("trace.localOnly must be true") }
            if let spanId = traceObj["spanId"] {
                guard let spanId = spanId as? String else { throw EnvelopeError("trace.spanId must be a string") }
                guard Ulid.isValid(spanId) else { throw EnvelopeError("trace.spanId must be a ULID") }
            }
            trace = traceObj
        }

        return Envelope(
            protocolId: protocolId,
            messageId: messageId,
            sessionId: sessionId,
            type: type,
            capability: capability,
            seq: seq,
            ackOf: ackOf,
            sentAtUnixMs: sentAtUnixMs,
            requiresAck: requiresAck,
            payload: payload,
            trace: trace
        )
    }

    private static func string(_ obj: [String: Any], _ field: String, pattern: String? = nil) throws -> String {
        guard let value = obj[field] as? String else {
            throw EnvelopeError("\(field) missing or not a string")
        }
        if let pattern, value.range(of: pattern, options: .regularExpression) == nil {
            throw EnvelopeError("bad \(field) '\(value)'")
        }
        return value
    }

    /// Booleans bridge to NSNumber too — require an integer-typed number.
    private static func integer(_ obj: [String: Any], _ field: String) throws -> Int64 {
        guard let number = obj[field] as? NSNumber, !isBoolean(number) else {
            throw EnvelopeError("\(field) missing or not a number")
        }
        if String(cString: number.objCType) == "d" {
            throw EnvelopeError("\(field) is not an integer")
        }
        return number.int64Value
    }

    private static func isBoolean(_ value: Any) -> Bool {
        CFGetTypeID(value as CFTypeRef) == CFBooleanGetTypeID()
    }
}
